larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,237 @@
+    const ResamplingPoints &resamplingPoints) :
+    m_xUpperBound(discreteProbabilityVector.m_xUpperBound),
+    m_useWidths(discreteProbabilityVector.m_useWidths),
+    m_discreteProbabilityData(this->ResampleDiscreteProbabilityData(discreteProbabilityVector, resamplingPoints))
+{
+    this->VerifyCompleteData();
+}
+
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float DiscreteProbabilityVector::EvaluateCumulativeProbability(const float x) const
+{
+    if (x - m_discreteProbabilityData.back().GetX() > std::numeric_limits<float>::epsilon())
+        return 1.f;
+
+    if (x - m_discreteProbabilityData.front().GetX() < std::numeric_limits<float>::epsilon())
+        return 0.f;
+
+    for (size_t iDatum = 1; iDatum < m_discreteProbabilityData.size(); ++iDatum)
+    {
+        if (x - m_discreteProbabilityData.at(iDatum).GetX() > std::numeric_limits<float>::epsilon())
+            continue;
+
+        const float xLow(m_discreteProbabilityData.at(iDatum-1).GetX());
+        const float yLow(m_discreteProbabilityData.at(iDatum-1).GetCumulativeDatum());
+        const float xHigh(m_discreteProbabilityData.at(iDatum).GetX());
+        const float yHigh(m_discreteProbabilityData.at(iDatum).GetCumulativeDatum());
+
+        if (std::fabs(xHigh-xLow) < std::numeric_limits<float>::epsilon())
+            throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+        const float m((yHigh-yLow)/(xHigh-xLow));
+        const float c(yLow-m*xLow);
+
+        return m*x+c;
+    }
+
+    throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename TX, typename TY>
+DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::InitialiseDiscreteProbabilityData(
+    InputData<TX, TY> inputData) const
+{
+    if (2 > inputData.size())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    std::sort(inputData.begin(), inputData.end(), DiscreteProbabilityVector::SortInputDataByX<TX,TY>);
+
+    if (inputData.back().first - m_xUpperBound > std::numeric_limits<float>::epsilon())
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    const float normalisation(this->CalculateNormalisation(inputData));
+    float accumulationDatum(0.f);
+
+    DiscreteProbabilityData data;
+    for (size_t iDatum = 0; iDatum < inputData.size()-1; ++iDatum)
+    {
+        const float x(static_cast<float>(inputData.at(iDatum).first));
+        const float deltaX(static_cast<float>(inputData.at(iDatum+1).first)-x);
+        const float densityDatum(static_cast<float>(inputData.at(iDatum).second) / normalisation);
+        accumulationDatum+=densityDatum*(m_useWidths ? deltaX : 1.f);
```

Comment:
Spaces around operators, throughout.

Reply:
Mentioned this above.

Mostly yes. When there are +- and * on the same line I tend to try to keep no spacing around * as it helps the mind employ bidmas.
I'll leave this one open if you want to come back to it.